genesyscloud/resource_genesyscloud_quality_forms_evaluation.go

```diff
@@ -0,0 +1,591 @@
+
+	// Publishing
+	if published {
+		_, _, err := qualityAPI.PostQualityPublishedformsEvaluations(platformclientv2.Publishform{
+			Id:        formId,
+			Published: &published,
+		})
+		if err != nil {
+			return diag.Errorf("Failed to publish evaluation form %s", name)
+		}
+	}
+
+	d.SetId(*formId)
+
+	log.Printf("Created evaluation form %s %s", name, *form.Id)
+	return readEvaluationForm(ctx, d, meta)
+}
+
+func readEvaluationForm(ctx context.Context, d *schema.ResourceData, meta interface{}) diag.Diagnostics {
+	sdkConfig := meta.(*providerMeta).ClientConfig
+	qualityAPI := platformclientv2.NewQualityApiWithConfig(sdkConfig)
+	log.Printf("Reading evaluation form %s", d.Id())
+
+	return withRetriesForRead(ctx, 30*time.Second, d, func() *resource.RetryError {
+		evaluationForm, resp, getErr := qualityAPI.GetQualityFormsEvaluation(d.Id())
+		if getErr != nil {
+			if isStatus404(resp) {
+				return resource.RetryableError(fmt.Errorf("Failed to read evaluation form %s: %s", d.Id(), getErr))
+			}
+			return resource.NonRetryableError(fmt.Errorf("Failed to read evaluation form %s: %s", d.Id(), getErr))
+		}
+
+		if evaluationForm.Name != nil {
+			d.Set("name", *evaluationForm.Name)
+		}
+		if evaluationForm.Published != nil {
+			d.Set("published", *evaluationForm.Published)
+		}
+		if evaluationForm.QuestionGroups != nil {
+			d.Set("question_groups", flattenQuestionGroups(evaluationForm.QuestionGroups))
+		}
+
+		return nil
+	})
+}
+
+func updateEvaluationForm(ctx context.Context, d *schema.ResourceData, meta interface{}) diag.Diagnostics {
+	name := d.Get("name").(string)
+	published := d.Get("published").(bool)
+
+	questionGroups, qgErr := buildSdkQuestionGroups(d)
+	if qgErr != nil {
+		return qgErr
+	}
+
+	sdkConfig := meta.(*providerMeta).ClientConfig
+	qualityAPI := platformclientv2.NewQualityApiWithConfig(sdkConfig)
+
+	// Get the latest unpublished version of the form
+	formVersions, _, err := qualityAPI.GetQualityFormsEvaluationVersions(d.Id(), 25, 1, "desc")
+	if err != nil {
+		return diag.Errorf("Failed to get evaluation form versions %s", name)
+	}
+
+	unpublishedForm := (*formVersions.Entities)[0]
+
+	log.Printf("Updating Evaluation Form %s", name)
+	form, _, err := qualityAPI.PutQualityFormsEvaluation(*unpublishedForm.Id, platformclientv2.Evaluationform{
+		Name:           &name,
+		QuestionGroups: questionGroups,
+	})
+	if err != nil {
+		return diag.Errorf("Failed to update evaluation form %s", name)
+	}
+
+	// Set published property on evaluation form update.
+	if published {
+		_, _, err := qualityAPI.PostQualityPublishedformsEvaluations(platformclientv2.Publishform{
+			Id:        form.Id,
+			Published: &published,
+		})
+		if err != nil {
+			return diag.Errorf("Failed to publish evaluation form %s", name)
+		}
+	} else {
+		// If published property is reset to false, set the resource Id to the latest unpublished form
+		d.SetId(*form.Id)
+	}
+
+	log.Printf("Updated evaluation form %s %s", name, *form.Id)
```

Comment:
It would be good to put in a sleep here of 5 seconds. The old form might be cached/or dynamo eventual consistency could cause the read to return stale data

Reply:
+ Added 5s sleep